tests/Application/Actions/Donations/UpdateTest.php

```diff
@@ -1765,4 +1754,242 @@ public function testAddDataSuccessWithCashBalanceAutoconfirm(): void
+        $donationInRepo = $this->getTestDonation();  // Get a new mock object so DB has old values.
+        // Make it explicit that the payment method type is (the unsupported
+        // for auto-confirms) "card".
+        $donationInRepo->setPaymentMethodType('card');
+
+        $donationRepoProphecy
+            ->findAndLockOneBy(['uuid' => '12345678-1234-1234-1234-1234567890ab'])
+            ->willReturn($donationInRepo)
+            ->shouldBeCalledOnce();
+        $donationRepoProphecy
+            ->push(Argument::type(Donation::class), false)
+            ->shouldNotBeCalled(); // Updates pushed to Salesforce
+
+        $entityManagerProphecy = $this->prophesize(EntityManagerInterface::class);
+        $entityManagerProphecy->beginTransaction()->shouldBeCalledOnce();
+        $entityManagerProphecy->rollback()->shouldBeCalledOnce();
+        $entityManagerProphecy->persist(Argument::type(Donation::class))->shouldNotBeCalled();
+        $entityManagerProphecy->flush()->shouldNotBeCalled();
+        $entityManagerProphecy->commit()->shouldNotBeCalled();
+
+        $stripePaymentIntentsProphecy = $this->prophesize(PaymentIntentService::class);
+        $stripePaymentIntentsProphecy->update(Argument::cetera())
+            ->shouldNotBeCalled();
+        $stripePaymentIntentsProphecy->confirm('pi_externalId_123')
+            ->shouldNotBeCalled();
+        $stripeClientProphecy = $this->prophesize(StripeClient::class);
+        $stripeClientProphecy->paymentIntents = $stripePaymentIntentsProphecy->reveal();
+
+        $container->set(DonationRepository::class, $donationRepoProphecy->reveal());
+        $container->set(EntityManagerInterface::class, $entityManagerProphecy->reveal());
+        $container->set(StripeClient::class, $stripeClientProphecy->reveal());
+
+        $requestPayload = $donation->toApiModel();
+        $requestPayload['autoConfirmFromCashBalance'] = true;
+        $request = $this->createRequest(
+            'PUT',
+            '/v1/donations/12345678-1234-1234-1234-1234567890ab',
+            json_encode($requestPayload),
+        )
+            ->withHeader('x-tbg-auth', DonationToken::create('12345678-1234-1234-1234-1234567890ab'));
+        $route = $this->getRouteWithDonationId('put', '12345678-1234-1234-1234-1234567890ab');
+
+        // act
+        $response = $app->handle($request->withAttribute('route', $route));
+
+        // assert
+        $payload = (string) $response->getBody();
+
+        $this->assertJson($payload);
+        $this->assertEquals(400, $response->getStatusCode());
+
+        $expectedPayload = new ActionPayload(400, ['error' => [
+            'type' => 'BAD_REQUEST',
+            'description' => 'Processing incomplete. Please refresh and check your donation funds balance',
+        ]]);
+        $expectedSerialised = json_encode($expectedPayload, JSON_PRETTY_PRINT);
+
+        $this->assertEquals($expectedSerialised, $payload);
```

Comment:
I see what this is doing, and maybe we discussed this before, but it feels like this test is quite long and fragile and it's a bit hard to see which code is the actual subject under test. As the business logic is mixed with the database stuff a bit in the Update controller it seems like the only way to test it actually works would be to do a test with a real DB. But I also think the controller is getting a bit long and unwieldy. We don't really have any service layer at the moment, but maybe it would be worth creating one now, and then we could have a much simpler test that just expects the service to throw an exception when there's a request to `autoConfirmFromCashBalance` on a card payment.